miio/device.py

```diff
@@ -122,205 +117,16 @@ def __init__(
-        with the device.
-
-        :rtype: Message
-
-        :raises DeviceException: if the device could not be discovered."""
-        m = Device.discover(self.ip)
-        if m is not None:
-            self._device_id = m.header.value.device_id
-            self._device_ts = m.header.value.ts
-            self._discovered = True
-            if self.debug > 1:
-                _LOGGER.debug(m)
-            _LOGGER.debug(
-                "Discovered %s with ts: %s, token: %s",
-                binascii.hexlify(self._device_id).decode(),
-                self._device_ts,
-                codecs.encode(m.checksum, "hex"),
-            )
-        else:
-            _LOGGER.error("Unable to discover a device at address %s", self.ip)
-            raise DeviceException("Unable to discover the device %s" % self.ip)
-
-        return m
-
-    @staticmethod
-    def discover(addr: str = None) -> Any:
-        """Scan for devices in the network.
-        This method is used to discover supported devices by sending a
-        handshake message to the broadcast address on port 54321.
-        If the target IP address is given, the handshake will be send as
-        an unicast packet.
-
-        :param str addr: Target IP address"""
-        timeout = 5
-        is_broadcast = addr is None
-        seen_addrs = []  # type: List[str]
-        if is_broadcast:
-            addr = "<broadcast>"
-            is_broadcast = True
-            _LOGGER.info("Sending discovery to %s with timeout of %ss..", addr, timeout)
-        # magic, length 32
-        helobytes = bytes.fromhex(
-            "21310020ffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
-        )
-
-        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
-        s.settimeout(timeout)
-        s.sendto(helobytes, (addr, 54321))
-        while True:
-            try:
-                data, addr = s.recvfrom(1024)
-                m = Message.parse(data)  # type: Message
-                _LOGGER.debug("Got a response: %s", m)
-                if not is_broadcast:
-                    return m
-
-                if addr[0] not in seen_addrs:
-                    _LOGGER.info(
-                        "  IP %s (ID: %s) - token: %s",
-                        addr[0],
-                        binascii.hexlify(m.header.value.device_id).decode(),
-                        codecs.encode(m.checksum, "hex"),
-                    )
-                    seen_addrs.append(addr[0])
-            except socket.timeout:
-                if is_broadcast:
-                    _LOGGER.info("Discovery done")
-                return  # ignore timeouts on discover
-            except Exception as ex:
-                _LOGGER.warning("error while reading discover results: %s", ex)
-                break
+        self.token = token
+        self.protocol = Protocol(ip, token, start_id, debug, lazy_discover)
```

Comment:
This could be `_protocol` to make it obvious that it's an internal and should not be accessed.

Reply:
✅ 

Had to expose `raw_id` property for `Vacuum` though.